Hub CA Certificate Migration

vilit1 opened the pull request:
Migrate the Certificate CA for IoT Hub from Baltimore to DigiCert.

Commands introduced:
- az iot hub certificate root-authority show
- az iot hub certificate root-authority set

Show has a default, since the service will return {} or None for that property. Set checks if the version chosen is already set before migration.

Note that this is a temporary command, will be removed after every hub has been migrated.

Note that this command will work only for some subscriptions and resource managers in the current state. Will update the pr as the feature becomes more avaliable.

---
This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.

Thank you for contributing to the IoT extension!

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

Intent for Production

- [ ] It is expected that pull requests made to default or core branches such as `dev` or `main` are of production grade. Corollary to this, any merged contributions to these branches may be deployed in a public release at any given time. By checking this box, you agree and commit to the expected production quality of code.

Basic expectations

- [ ] If introducing new functionality or modified behavior, are they backed by unit and/or integration tests?
- [ ] In the same context as above are command names and their parameter definitions accurate? Do help docs have sufficient content?
- [ ] Have **all** the relevant unit **and** integration tests pass? i.e. `pytest <project root> -vv`. Please provide evidence in the form of a screenshot showing a succesful run of tests locally OR a link to a test pipeline that has been run against the change-set.
- [ ] Have linter checks passed using the `.pylintrc` and `.flake8` rules? Look at the CI scripts for example usage.
- [ ] Have extraneous print or debug statements, commented out code-blocks or code-statements (if any) been removed from the surface area of changes?
- [ ] Have you made an entry in HISTORY.rst which concisely explains your user-facing feature or change?

Azure IoT CLI maintainers reserve the right to enforce any of the outlined expectations.

A PR is considered **ready for review** when all basic expectations have been met (or do not apply).

avagraw commented on azext_iot/iothub/params.py:
```diff
@@ -246,3 +247,11 @@ def load_iothub_arguments(self, _):
             help="MIME Type of file.",
         )
+
+    with self.argument_context("iot hub certificate root-authority set") as context:
```
The -n parameter for iothub_name needs to be added

vilit1 replied:
it is inherited from "iot" - https://github.com/Azure/azure-iot-cli-extension/blob/73f5ff3cb283bb8106db2dd81b126b8d25c24b1a/azext_iot/_params.py#L95

So this shouldn't be breaking the pylint - otherwise other commands within the iothub folder that use hub name would also trigger this.

vilit1 replied:
Adding this in - there may be scope mismatch since this command may inherit from iot hub certificate

avagraw commented on azext_iot/iothub/providers/certificate.py:
```diff
@@ -0,0 +1,107 @@
+        current_target = self._get_target_root_certificate()
+        if not current_target:
+            return
+        root_ca = current_target.get("enableRootCertificateV2") or False
+
+        # Check if changes are needed
+        if (
+            (root_ca and ca_version == CertificateAuthorityVersions.v2.value) or
+            (not root_ca and ca_version == CertificateAuthorityVersions.v1.value)
+        ):
+            print(NO_CHANGE_MSG.format(ca_version))
+            return
+        elif ca_version == CertificateAuthorityVersions.v2.value:
+            print(CA_TRANSITION_WARNING)
+            if not yes and not prompt_y_n(msg=CONT_INPUT_MSG, default="n"):
+                raise ManualInterrupt(ABORT_MSG)
+            root_ca = True
+        else:
+            print(CA_REVERT_WARNING)
+            if not yes and not prompt_y_n(msg=CONT_INPUT_MSG, default="n"):
+                raise ManualInterrupt(ABORT_MSG)
+            root_ca = False
```
This can be simplified:
```
is_root_ca = current_target.get("enableRootCertificateV2") or False
if (ca_version == CertificateAuthorityVersions.v1.value and is_root_ca):
    print(CA_REVERT_WARNING)
    if not yes and not prompt_y_n(msg=CONT_INPUT_MSG, default="n"):
        raise ManualInterrupt(ABORT_MSG)
elif (ca_version == CertificateAuthorityVersions.v2.value and not is_root_ca):
    print(CA_TRANSITION_WARNING)
    if not yes and not prompt_y_n(msg=CONT_INPUT_MSG, default="n"):
        raise ManualInterrupt(ABORT_MSG)
else:
    print(NO_CHANGE_MSG.format(ca_version))
    return
enable_root_ca = not is_root_ca
```
Then pass in the `enable_root_ca` when running resource update command.

avagraw commented:
Please post the pipeline run showing the testing evidence. 

@digimaun Since, this is a temporary change, should we update history?

digimaun commented on azext_iot/tests/iothub/certificate/test_iothub_certificate_root_authority_int.py:
```diff
@@ -0,0 +1,67 @@
+        super(TestIotHubCertificateRoot, self).__init__(test_case)
+
+    def test_certificate_root(self):
+        for auth_phase in DATAPLANE_AUTH_TYPES:
```
There are no connection string or keys at play for this control plane only functionality.

digimaun commented on azext_iot/iothub/providers/certificate.py:
```diff
@@ -0,0 +1,102 @@
+        cmd,
+        hub_name: Optional[str] = None,
+        rg: Optional[str] = None,
+        login: Optional[str] = None,
```
Why support login or auth_type_dataplane on init here? I was under the impression this was control plane only functionality.

vilit1 replied:
tbh this was more for the discovery of the hub rather than actually calling the command (and me not knowing better back then). I will remove cause it is more or less pointless

digimaun commented on azext_iot/iothub/commands_certificate.py:
```diff
@@ -0,0 +1,50 @@
+    resource_group_name: Optional[str] = None,
+    login: Optional[str] = None,
+    auth_type_dataplane: Optional[str] = None,
+    yes: Optional[str] = False
```
Optional typing means the explicit value of None is supported. Also, what is here is Optional typing of a str, where-as the default assigned value is bool and the iot_hub_certificate_root_authority_set type hint is straight bool

![image](https://user-images.githubusercontent.com/28658112/186040283-96780dd0-d727-42d2-8fe5-25c4e080ca25.png)

So is this typing declaration the most accurate usage?

vilit1 replied:
typo

digimaun commented on azext_iot/iothub/commands_certificate.py:
```diff
@@ -0,0 +1,50 @@
+    cmd,
+    hub_name: Optional[str] = None,
+    resource_group_name: Optional[str] = None,
+    login: Optional[str] = None,
```
if this is control plane only functionality, `login` and `auth_type_dataplane` should be removed as function params.

digimaun commented on azext_iot/iothub/_help.py:
```diff
@@ -309,3 +309,33 @@ def load_iothub_help():
+        long-summary: Transition this resource to a certificate on the DigiCert Global G2 root. Before making this
+          transition, please ensure all devices are updated to contain the public portion of the G2 root. Devices
+          will disconnect and reconnect using the new root.
+          We suggest monitoring current connections but another metric may be more appropriate for your situation.
```
What is another metric this last sentence is referring to?

vilit1 replied:
not sure - these strings are copy paste from the portal.

The portal includes a link that directs to the metrics page but it doesn't auto populate the metrics so "another metric" may just be something the customer defines and monitors themselves.

vilit1 replied:
redefined to user defined metric to be less awk

digimaun commented on azext_iot/iothub/providers/certificate.py:
```diff
@@ -0,0 +1,102 @@
+        if not result.success():
+            # Error will already be printed out
+            return
+        current_target = result.as_json().get("properties").get("rootCertificate")
```
tiny nit, if 'properties' will always be there, you might as well do `result.as_json()["properties"].get("rootCertificate")`. If for reasons, there is a _chance_ a properties object/dict won't be returned then we can be slightly more defensive by doing `result.as_json().get("properties", {}).get("rootCertificate")`  (providing the default dict fallback on the first get).

digimaun commented:
Please add an entry to `HISTORY.rst` under `0.17.1` and we can increment the version of the extension in this PR to `0.17.1` 

https://github.com/Azure/azure-iot-cli-extension/blob/dev/azext_iot/constants.py#L10

vilit1 commented:
Durations with a _**pre-created**_ IoT Hub with the correct feature enabled:
![image](https://user-images.githubusercontent.com/73560279/186535575-25d992fb-826c-4e2b-9a41-65829c109240.png)

digimaun commented:
> Please post the pipeline run showing the testing evidence.
> 
> @digimaun Since, this is a temporary change, should we update history?

I would still because its user facing, even if temporary, in fact our release notes can use some kind of "limited period availability" nomenclature.

digimaun commented on HISTORY.rst:
```diff
@@ -7,6 +7,18 @@ unreleased
+
+**IoT Hub Update**
+
+* The root-authority migration feature is now avaliable. Since the Baltimore root will soon expire, IoT Hub will
```
typo avaliable

digimaun commented on azext_iot/iothub/_help.py:
```diff
@@ -309,3 +309,33 @@ def load_iothub_help():
+    helps["iot hub certificate root-authority set"] = """
+        type: command
+        short-summary: Set the certificate root-authority for an IoT Hub instance to a specific version.
+        long-summary: Transition this resource to a certificate on the DigiCert Global G2 root or revert to Baltimore root.
```
tiny nit, maybe we can add in parenthesis, the connection between a root certificate description and v2 or v1 i.e. `Transition this resource to a certificate on the DigiCert Global G2 root (v2) or revert to Baltimore root (v1).`

vilit1 commented:
Added a fix for the initial root authority change

tests passing (local run, on a different sub, **still in canary env**):
![image](https://user-images.githubusercontent.com/73560279/187761911-987354ab-48bd-4261-8388-03e742582961.png)

Note that for these tests, I modified the tests so that:
1. the rbac "IoT Hub Data Contributor" role is not assigned (I do not have permissions in the test sub, and is not needed for this test)
2. the dataplane actions (check for and delete any devices, configurations) during teardown are skipped. I will add back the dataplane actions once there is a fix on the service side.

digimaun commented on HISTORY.rst:
```diff
@@ -6,6 +6,17 @@ Release History
 unreleased
 +++++++++++++++
 
+**IoT Hub Update**
```
Lets put these notes under 0.18.1 since that is going out next, also tiny nit what about "**IoT Hub updates**" for the header?

digimaun commented on azext_iot/iothub/commands_certificate.py:
```diff
@@ -0,0 +1,42 @@
+
+def certificate_root_authority_show(
+    cmd,
+    hub_name: Optional[str] = None,
```
Is hub_name really Optional and should it have a default value of None?

digimaun commented on azext_iot/iothub/commands_certificate.py:
```diff
@@ -0,0 +1,42 @@
+def certificate_root_authority_set(
+    cmd,
+    ca_version: str,
+    hub_name: Optional[str] = None,
```
Same comment as above for hub_name

digimaun commented on azext_iot/iothub/providers/certificate.py:
```diff
@@ -0,0 +1,98 @@
+
+        # Check if changes are needed
+        if ca_version == CertificateAuthorityVersions.v1.value and root_ca:
+            print(CA_REVERT_WARNING)
```
Why not use logger.warning?

digimaun commented on azext_iot/iothub/providers/certificate.py:
```diff
@@ -0,0 +1,98 @@
+            if not yes and not prompt_y_n(msg=CONT_INPUT_MSG, default="n"):
+                raise ManualInterrupt(ABORT_MSG)
+        elif ca_version == CertificateAuthorityVersions.v2.value and not root_ca:
+            print(CA_TRANSITION_WARNING)
```
Why not use logger.warning?

digimaun commented on azext_iot/iothub/providers/certificate.py:
```diff
@@ -0,0 +1,98 @@
+    def __init__(
+        self,
+        cmd,
+        hub_name: Optional[str] = None,
```
Same hub_name comment as earlier

vilit1 commented:
![image](https://user-images.githubusercontent.com/73560279/200663031-a2d88938-3645-48f7-a7f1-dc00cf5d587b.png)

digimaun commented on azext_iot/iothub/providers/certificate.py:
```diff
@@ -0,0 +1,98 @@
+            if not yes and not prompt_y_n(msg=CONT_INPUT_MSG, default="n"):
+                raise ManualInterrupt(ABORT_MSG)
+        else:
+            print(NO_CHANGE_MSG.format(ca_version))
```
Why not use logger.warning?

digimaun commented on azext_iot/iothub/providers/certificate.py:
```diff
@@ -0,0 +1,98 @@
+
+        command = "resource update -n {} -g {} --api-version {} --resource-type {}".format(
+            self.target["entity"].split(".")[0],
+            self.target['resourcegroup'],
```
nit, consistency with single quote vs double quote

digimaun commented on azext_iot/tests/iothub/__init__.py:
```diff
@@ -50,10 +50,11 @@
         assert test_scenario
         self.entity_rg = ENTITY_RG
         self.entity_name = ENTITY_NAME
+        self.added_data_contributor = add_data_contributor
```
does `self.added_data_contributor` get used anywhere?

digimaun commented:
LGTM pending bash feedback, and successful IT run. +We discussed RBAC simplification strategies for tests offline.

vilit1 commented:
https://dev.azure.com/azureiotdevxp/aziotcli/_build/results?buildId=6956&view=results

vilit1 commented:
https://dev.azure.com/azureiotdevxp/aziotcli/_build/results?buildId=7077&view=results
